### PR TITLE
Swap slippage tolerance

### PR DESCRIPTION
**Pull request by gromxyz**
- Extend `Swap` component to include a `slippage tolerance` component (icon gear + select box)
- Store selected `slippage tolerance` in `src/renderer/services/app/service.ts`

Part of #1105

### CODE REVIEW

File: src/renderer/components/currency/CurrencyInfo.styles.ts [hunk 1]
```diff
@@ -29,3 +34,26 @@ export const Container = styled('div')`
...
+
+export const DropdownContentWrapper = styled(Row)`
+  justify-content: space-between;
+  padding-right: 0;
+  align-items: center;
+  width: 100%;
+  cursor: pointer;
+`
+
+export const SlipLabel = styled(Text)<{ slip: String }>`
```
Comment (veado) [inline src/renderer/components/currency/CurrencyInfo.styles.ts]: No need to define `slip` props (`<{ slip: String }>`) here, it's not used in any style.

File: src/renderer/components/currency/CurrencyInfo.styles.ts [hunk 2]
```diff
@@ -29,3 +34,26 @@ export const Container = styled('div')`
...
+  text-transform: uppercase;
+  padding: 0;
+  font-size: 16px;
+  font-family: 'MainFontRegular';
+`
+export const SlipSettings = styled(SlipSettingsIcon)<{ active: true | false }>`
+  & path {
+    fill: ${({ active }) => (active === true ? palette('text', 2) : palette('gray', 2))};
```
Comment (veado) [inline src/renderer/components/currency/CurrencyInfo.styles.ts]: [nitpick] `active === true ? ...` _> `active ? ...`

File: src/renderer/components/currency/CurrencyInfo.tsx [hunk 3]
```diff
@@ -1,21 +1,67 @@
...
+        ))}
+      </>
+    )
+  }, [changeSlipToleranceHandler])
+
+  const renderSlipSettings = useMemo(
+    () => (
+      <Dropdown overlay={slipSettings} trigger={['click']} placement="bottomCenter">
+        <Styled.DropdownContentWrapper>
+          <Row style={{ alignItems: 'center' }}>
```
Comment (veado) [inline src/renderer/components/currency/CurrencyInfo.tsx]: `DropdownContentWrapper` is already a `Row`, no need to put another `Row` as child. Add `align-items: center` to `DropdownContentWrapper` in `src/renderer/components/currency/CurrencyInfo.styles.ts`

File: src/renderer/components/currency/CurrencyInfo.tsx [hunk 4]
```diff
@@ -1,21 +1,67 @@
...
+            <Styled.SlipLabel slip={`${slipTolerance}%`}>{slipTolerance}%</Styled.SlipLabel>
```
Comment (veado) [inline src/renderer/components/currency/CurrencyInfo.tsx]: No need to have `slip` property - see https://github.com/thorchain/asgardex-electron/pull/1635#discussion_r670433883

File: src/renderer/assets/svg/icon-slip-settings.svg [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<svg width="18" height="18" viewBox="0 0 18 18" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (veado) [inline src/renderer/assets/svg/icon-slip-settings.svg]: Can't we use `icon-settings.svg` we already have?

File: src/renderer/components/currency/CurrencyInfo.tsx [hunk 6]
```diff
@@ -1,21 +1,67 @@
...
 type CurrencyInfoProps = {
   from?: O.Option<PoolAssetDetail>
   to?: O.Option<PoolAssetDetail>
   slip?: BigNumber
+  slipTolerance: SlipTolerance
+  changeSlipTolerance: ChangeSlipToleranceHandler
 }
 
-export const CurrencyInfo = ({ to = O.none, from = O.none, slip = bn(0) }: CurrencyInfoProps) => {
+const SLIP_PERCENTAGES = [3, 5, 10]
```
Comment (veado) [inline src/renderer/components/currency/CurrencyInfo.tsx]: [nitpick] Type array elements as `SlipTolerance`:

 ```ts
 const SLIP_PERCENTAGES: SlipTolerance[] = [...]
 ```

File: src/renderer/components/currency/CurrencyInfo.tsx [hunk 7]
```diff
@@ -1,21 +1,67 @@
...
+
+export const CurrencyInfo = ({ to = O.none, from = O.none, slipTolerance, changeSlipTolerance }: CurrencyInfoProps) => {
+  const [slipSettingsVisible, setSlipSettingsVisible] = useState(false)
+
+  const changeSlipToleranceHandler = useCallback(
+    (slipTolerance) => {
+      changeSlipTolerance(slipTolerance as SlipTolerance)
```
Comment (veado) [inline src/renderer/components/currency/CurrencyInfo.tsx]: Then no need to type cast `... as SlipTolerance` I think... See https://github.com/thorchain/asgardex-electron/pull/1635#discussion_r670440613

File: src/renderer/components/currency/CurrencyInfo.tsx [hunk 8]
```diff
@@ -1,21 +1,67 @@
...
+    },
+    [changeSlipTolerance]
+  )
+
+  const slipSettings = useMemo(() => {
+    return (
+      <>
+        {SLIP_PERCENTAGES.map((slip) => (
+          <Row style={{ alignItems: 'center' }} key={slip}>
+            <Styled.SlipLabel key={slip} onClick={() => changeSlipToleranceHandler(slip)} slip={`${slip}%`}>
```
Comment (veado) [inline src/renderer/components/currency/CurrencyInfo.tsx]: `slip`property can be removed - See https://github.com/thorchain/asgardex-electron/pull/1635#discussion_r670433883

File: src/renderer/components/currency/CurrencyInfo.tsx [hunk 9]
```diff
@@ -47,7 +93,21 @@ export const CurrencyInfo = ({ to = O.none, from = O.none, slip = bn(0) }: Curre
...
+            <div style={{ display: 'flex', alignItems: 'center' }}>
+              slip: {slipTolerance}%&nbsp;
+              <Styled.SlipSettings
+                active={slipSettingsVisible}
+                onClick={() => setSlipSettingsVisible(!slipSettingsVisible)}
+              />
+            </div>
+            {slipSettingsVisible && (
+              <div>
+                <Styled.SlipToleranceText>slippage tolerance</Styled.SlipToleranceText>
```
Comment (veado) [inline src/renderer/components/currency/CurrencyInfo.tsx]: Use `i18n` for `slippage tolerance`.

File: src/renderer/components/currency/CurrencyInfo.tsx [hunk 10]
```diff
@@ -47,7 +93,21 @@ export const CurrencyInfo = ({ to = O.none, from = O.none, slip = bn(0) }: Curre
               trimZeros: true
             })}
           </div>
-          <div>slip: {slip.multipliedBy(100).toFormat(2)}%</div>
+          <div>
+            <div style={{ display: 'flex', alignItems: 'center' }}>
+              slip: {slipTolerance}%&nbsp;
```
Comment (veado) [inline src/renderer/components/currency/CurrencyInfo.tsx]: Use `i18n` for `slip`.

Comment (veado) [reply src/renderer/components/currency/CurrencyInfo.tsx]: Can you try to add some space via `css` (not by using `&nbsp;`)?

**Review by veado — CHANGES_REQUESTED** [review]
Good work! Some comments have been added (just some "fine-tuning").

File: src/renderer/i18n/de/swap.ts [hunk 11]
```diff
@@ -6,6 +6,8 @@ const swap: SwapMessages = {
   'swap.state.success': 'Erfolgreich getauscht',
   'swap.input': 'Eingabe',
   'swap.output': 'Ausgabe',
+  'swap.slip.title': 'Schlupf',
+  'swap.slip.tolerance': 'Schlupftoleranz',
```
Comment (veado) [inline src/renderer/i18n/de/swap.ts]: 'swap.slip.title': 'Slip',
  'swap.slip.tolerance': 'Sliptoleranz',

File: src/renderer/i18n/fr/swap.ts [hunk 12]
```diff
@@ -6,6 +6,8 @@ const swap: SwapMessages = {
   'swap.state.error': "Erreur lors de l'échange",
   'swap.input': 'Entrée',
   'swap.output': 'Sortie',
+  'swap.slip.title': 'Glissement',
+  'swap.slip.tolerance': 'Tolérance au glissement',
```
Comment (veado) [inline src/renderer/i18n/fr/swap.ts]:
```ts
  'swap.slip.title': 'Slip - FR',
  'swap.slip.tolerance': 'Slippage tolerance - FR',
  ```

File: src/renderer/i18n/ru/swap.ts [hunk 13]
```diff
@@ -6,6 +6,8 @@ const swap: SwapMessages = {
   'swap.state.error': 'Ошибка при обмене',
   'swap.input': 'Отдаете',
   'swap.output': 'Получаете',
+  'swap.slip.title': 'Проскальзывание',
+  'swap.slip.tolerance': 'Допуск по проскальзыванию',
```
Comment (veado) [inline src/renderer/i18n/ru/swap.ts:10]:
```ts
  'swap.slip.title': 'Slip - RU',
  'swap.slip.tolerance': 'Slippage tolerance - RU',
  ```

File: src/renderer/views/swap/SwapView.tsx [hunk 14]
```diff
@@ -168,6 +168,20 @@ export const SwapView: React.FC<Props> = (_): JSX.Element => {
     reloadInboundAddresses()
   }, [reloadBalances, reloadInboundAddresses, reloadPools])
 
+  const getStoredSlipTolerance = (): SlipTolerance =>
+    FP.pipe(
+      localStorage.getItem(SLIP_TOLERANCE_KEY) as string,
```
Comment (veado) [inline src/renderer/views/swap/SwapView.tsx]: `as string` can be removed

File: src/renderer/views/swap/SwapView.tsx [hunk 15]
```diff
@@ -168,6 +168,20 @@ export const SwapView: React.FC<Props> = (_): JSX.Element => {
...
+      O.fromNullable,
+      O.map((s) => {
+        const slipTolerance = +s as SlipTolerance
+        changeSlipTolerance(slipTolerance)
+        return slipTolerance
```
Comment (veado) [inline src/renderer/views/swap/SwapView.tsx]: How about the following to have stronger types? 

1. Define a type guard `isSlipTolerance` in `src/shared/api/types.ts`

```ts
export const isSlipTolerance = (value: SlipTolerance | number): value is SlipTolerance => {
  const t: SlipTolerance[] = [3, 5, 10]
  return t.includes(value as SlipTolerance)
}
``` 

2. to use it here

```ts
      O.map((s) => {
        const itemAsInt = parseInt(s)
        const slipTolerance = isSlipTolerance(itemAsInt) ? itemAsInt : DEFAULT_SLIP_TOLERANCE
        changeSlipTolerance(slipTolerance)
        return slipTolerance
      }),
```

**Review by veado — CHANGES_REQUESTED** [review]
Looks nice! Some comments for `i18n` and types.

File: src/renderer/views/swap/SwapView.tsx (hunk 15 above)
Comment (veado) [reply src/renderer/views/swap/SwapView.tsx]: BTW: Pls move `SlipTolerance` + `isSlipTolerance` from (1) `shared/api/types` to (2) `src/renderer/types/asgardex.ts`. We use (1) for sharing code in `main` and `renderer` processes in Electron, but those types are needed for `renderer` only

Comment (gromxyz) [reply src/renderer/views/swap/SwapView.tsx]: Makes sense, changes are on the way

**Pull request merged**